extensions/amp-ad-exit/0.1/test-e2e/test-amp-ad-exit.js

```diff
@@ -0,0 +1,96 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+describes.endtoend(
+  'amp-ad-exit',
+  {
+    testUrl:
+      'http://localhost:8000/test/fixtures/e2e/amphtml-ads/amp-ad-exit.amp.html',
+    environments: 'amp4ads-preset',
+  },
+  env => {
+    let controller;
+
+    beforeEach(async () => {
+      controller = env.controller;
+    });
+
+    // Setting the time explicitly to avoid test flakiness.
+    async function setTime(epochTime) {
+      await controller.evaluate(time => {
+        window.Date = {
+          now: () => time,
+        };
+        if (window.parent) {
+          try {
+            window.parent.Date = window.Date;
+          } catch (e) {}
+        }
+      }, epochTime);
+    }
+
+    it('product1 clicked without delay', async () => {
+      const adDiv = await controller.findElement('#product1');
+      await controller.click(adDiv);
+
+      const windows = await controller.getAllWindows();
+      await expect(windows.length).to.equal(1);
+    });
+
+    it('product1 clicked near the border', async () => {
+      const adDiv = await controller.findElement('#product1');
+      await setTime(Number.MAX_VALUE);
```

Comment:
What is `setTime()`? Do you mean `setTimeout()`?

Reply:
`setTime` sets the time `Date.now()` would return. Certain logic being tested here is time-based and I don't want e2e to depend on the real timer, so I did this here.